### PR TITLE
[executor]: Add runtime filter properties

### PR DESCRIPTION
**Pull request by elsa0520**
### What problem does this PR solve?

Issue Number: close #https://github.com/pingcap/tidb/issues/40220

Problem Summary:
Add RuntimeFilter in TableScan and HashJoinNode

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

How it Works:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### CODE REVIEW

**Review by winoros — COMMENTED** [review]
the core pb change lgtm

File: proto/executor.proto [hunk 1]
```diff
@@ -116,6 +118,7 @@ message PartitionTableScan {
 	repeated int64 partition_ids = 6;
 	optional bool is_fast_scan = 7; // fast_scan is a feature only provided by TiFlash (but not TiKV).
 	repeated Expr pushed_down_filter_conditions = 8; // conditions that are pushed down to storage layer, only used by TiFlash.
+	repeated RuntimeFilter runtime_filter_list = 9; // only used by TiFlash
```
Comment (windtalker) [inline proto/executor.proto:121]: Why don't have `max_wait_time_ms` for partition table scan?

File: proto/executor.proto [hunk 2]
```diff
@@ -147,6 +150,27 @@ message Join {
...
 	optional bool is_null_aware_semi_join = 13;
+	repeated RuntimeFilter runtime_filter_list = 14; // only used by TiFlash
+}
+
+message RuntimeFilter {
+	optional int32 id = 1 [(gogoproto.nullable) = false];
+	optional Expr source_expr = 2;
+	optional Expr target_expr = 3;
+	optional string source_executor_id = 4 [(gogoproto.nullable) = false];
+	optional string target_executor_id = 5 [(gogoproto.nullable) = false];
```
Comment (windtalker) [inline proto/executor.proto:161]: Can you give an example to explain these fields?

File: proto/executor.proto [hunk 3]
```diff
@@ -147,6 +150,27 @@ message Join {
...
+	optional Expr rf_expr = 6;
+	optional RuntimeFilterType rf_type = 7 [(gogoproto.nullable) = false];
+	optional RuntimeFilterMode rf_mode = 8 [(gogoproto.nullable) = false];
+}
+
+enum RuntimeFilterType {
+	IN = 0;
+	MIN_MAX = 1;
```
Comment (windtalker) [inline proto/executor.proto:169]: Maybe we can add all the potential types that we may support, so next time when supporting a new RuntimeFilterType, there is no need to update the protocol?

File: proto/executor.proto [hunk 4]
```diff
@@ -147,6 +150,27 @@ message Join {
...
+}
+
+enum RuntimeFilterMode {
+	LOCAL = 0;
```
Comment (windtalker) [inline proto/executor.proto:173]: Ditto

File: proto/executor.proto [hunk 5]
```diff
@@ -163,6 +187,7 @@ message Selection {
 	repeated Expr conditions = 1;
 	repeated RpnExpr rpn_conditions = 2;
 	optional Executor child = 3;
+	optional bool has_rf_conditions = 4 [(gogoproto.nullable) = false];
```
Comment (windtalker) [inline proto/executor.proto:190]: Will `rf_condition` be put in `Selection`?

File: proto/executor.proto (hunk 2 above)
Comment (elsa0520) [reply proto/executor.proto:161]: I will explain it in planner code

File: proto/executor.proto (hunk 5 above)
Comment (elsa0520) [reply proto/executor.proto]: Remove it.

**Review by windtalker — DISMISSED** [review]
LGTM

**Review by winoros — APPROVED** [review]

File: proto/executor.proto [hunk 6]
```diff
@@ -116,6 +118,8 @@ message PartitionTableScan {
 	repeated int64 partition_ids = 6;
 	optional bool is_fast_scan = 7; // fast_scan is a feature only provided by TiFlash (but not TiKV).
 	repeated Expr pushed_down_filter_conditions = 8; // conditions that are pushed down to storage layer, only used by TiFlash.
+	repeated RuntimeFilter runtime_filter_list = 9; // only used by TiFlash
+        optional int32 max_wait_time_ms = 10  [(gogoproto.nullable) = false];;
```
Comment (AilinKid) [inline proto/executor.proto:122]: multi ;;?

**Pull request merged**